libvmi/driver/kvm/kvm_private.h

```diff
@@ -30,18 +30,34 @@
 
 #include <libvirt/libvirt.h>
 #include <libvirt/virterror.h>
+#ifndef ENABLE_KVM_LEGACY
+# include <kvmi/libkvmi.h>
+#endif
+
 
 #include "private.h"
 #include "libvirt_wrapper.h"
+#include "libkvmi_wrapper.h"
 
 typedef struct kvm_instance {
     virConnectPtr conn;
     virDomainPtr dom;
     uint32_t id;
     char *name;
     char *ds_path;
-    int socket_fd;
     libvirt_wrapper_t libvirt;
+#ifdef ENABLE_KVM_LEGACY
+    int socket_fd;
+#else
+    void *kvmi;
+    void *kvmi_dom;
+    libkvmi_wrapper_t libkvmi;
+    pthread_mutex_t kvm_connect_mutex;
+    pthread_cond_t kvm_start_cond;
+    unsigned int expected_pause_count;
+    // dispatcher to handle VM events in each process_xxx functions
+    status_t (*process_event[KVMI_NUM_EVENTS])(vmi_instance_t vmi, struct kvmi_dom_event *event);
+#endif
```

Comment:
So instead of this ifdeffery, it would be better to split the legacy driver and the new driver completely. Why do you want to rule out the possibility of building a single libvmi instance that can work with both?

Reply:
So adding a `KVM_LEGACY` driver is doable, I just didn't think we would keep both at the same time.

Reply:
I don't know, I think it might make sense to just have a single build of LibVMI that can target a bunch of different setups. I mean, that's kinda the point of LibVMI :)

Reply:
> I mean, that's kinda the point of LibVMI :)

Well yes, after all.
But I would rather keep this refactoring for another PR, as this one is really huge already.

Reply:
Fair enough.